src_features/signMessageEIP712/typed_data.c

```diff
@@ -0,0 +1,749 @@
+#ifdef HAVE_EIP712_FULL_SUPPORT
+
+#include <stdlib.h>
+#include <string.h>
+#include "typed_data.h"
+#include "sol_typenames.h"
+#include "apdu_constants.h"  // APDU response codes
+#include "context_712.h"
+#include "mem.h"
+#include "mem_utils.h"
+
+static s_typed_data *typed_data = NULL;
+
+/**
+ * Initialize the typed data context
+ *
+ * @return whether the memory allocation was successful
+ */
+bool typed_data_init(void) {
+    if (typed_data == NULL) {
+        if ((typed_data = MEM_ALLOC_AND_ALIGN_TYPE(*typed_data)) == NULL) {
+            apdu_response_code = APDU_RESPONSE_INSUFFICIENT_MEMORY;
+            return false;
+        }
+        // set types pointer
+        if ((typed_data->structs_array = mem_alloc(sizeof(uint8_t))) == NULL) {
+            apdu_response_code = APDU_RESPONSE_INSUFFICIENT_MEMORY;
+            return false;
+        }
+
+        // create len(types)
+        *(typed_data->structs_array) = 0;
+    }
+    return true;
+}
+
+void typed_data_deinit(void) {
+    typed_data = NULL;
```

Comment:
Since this module is not the only one using the memory allocator, it cannot on its own using dealloc in the deinit function since it might dealloc something else (as the mem lib works as a simple stack).
This is the job of the function that calls typed_data_deinit() and there is indeed a mem_reset() in the eip712_context_deinit function after typed_data_deinit() is called.